### ISSUE TITLE
[Feature] Implement functions definitions, declarations and call expressions for user-defined functions

### ISSUE THREAD

**Issue by Luna-Klatzer** [Kipper-Lang/Kipper#183]
### Is there an existing proposal for this?

- [X] I have searched the existing issues

### This feature does not exist in the latest version

- [X] I am using the latest version

### Proposal

Implement ability to declare and define functions, and call them using function-call expressions (`call IDENTIFIER(ARGS...)`):

- Syntax for declarations:
  ```ts
  def NAME(ARG1: TYPE1, ARG2: TYPE2, ARG3: TYPE3, ..., ARGn: TYPEn) -> RETURN_TYPE;
  ```
- Syntax for definitions:
  ```ts
  def NAME(ARG1: TYPE1, ARG2: TYPE2, ARG3: TYPE3, ..., ARGn: TYPEn) -> RETURN_TYPE {
    STATEMENTS... // Optional - Function content

    return; // Optional if RETURN_TYPE is void, invalid if RETURN_TYPE is not void
    return VALUE; // Required if RETURN_TYPE is not void
  }
  ```
- Function call expressions with user-defined arguments:
  ```ts
  // Function definition with specific arguments and return type
  def func(arg1: T1, arg2: T2, arg3: T3, ..., argn: Tn) -> RetT { ... }

  // Call expression
  var ret: RetT = call func(arg1, arg2, arg3, ..., argn);
  ``` 

### Exact behaviour / changes you want

- [x] Implement syntax for function declarations and definitions.
- [ ] Implement semantics and type checking for function definitions and call expressions.
- [ ] Implement function-specific return statement and type checking for return types.
- [ ] Add support for local-scope non-global functions.
- [ ] Update implementation of `ScopeFunctionDeclaration` to support local-scope functions.
- [ ] Remove ability to declare functions without a function body.
- [ ] Update docs page and add more examples.

**Comment by Luna-Klatzer** [comment]
Beginning with d6d2887ca2bc39c942a691f8edaf18fba8ee13e1 support for return statements and basic return type checking are mostly complete. With previous commits also function referencing, function call arguments and function argument type checking have been implemented.

Remaining tasks:
- [x] Branch checking to avoid functions not having a return (unless their return is `void`)
- [x] Argument referencing and type checking.
- [x] Function scope proper checking and compatibility with arguments.
- [x] Tests for function arguments, return type-checking and use of function returns in an expression.

**Comment by Luna-Klatzer** [comment]
Implemented with #325 and released with `v0.10.0-alpha.5` (#329)
